Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -1541,13 +1541,41 @@ extension BrowserViewController: URLBarDelegate {
     }
 
     func urlBar(urlBar: URLBarView, didSubmitText text: String) {
-        // If we can't make a valid URL, do a search query.
-        // If we still don't have a valid URL, something is broken. Give up.
+        let trimmedText = text.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceCharacterSet())
+
+        guard let possibleKeywordQuerySeparatorSpace = trimmedText.characters.indexOf(" ") else {
+            submitText(text)
+            return
+        }
+
+        let possibleKeyword = trimmedText.substringToIndex(possibleKeywordQuerySeparatorSpace)
+        let possibleQuery = trimmedText.substringFromIndex(possibleKeywordQuerySeparatorSpace.successor())
+
+        profile.bookmarks.getURLForKeywordSearch(possibleKeyword).uponQueue(dispatch_get_main_queue()) { result in
+            if var urlString = result.successValue,
+                let escapedQuery = possibleQuery.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.SearchTermsAllowedCharacterSet()),
+                let range = urlString.rangeOfString("%s") {
+                urlString.replaceRange(range, with: escapedQuery)
+
+                if let url = NSURL(string: urlString) {
+                    let engine = self.profile.searchEngines.defaultEngine
+                    Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
```

Comment:
@thebnich, removed those 2 lines. Also updated the order in which the parsing of the text happens. If it's a URL, I felt it would make sense to submit it immediately. Rather than wait for keyword lookup.